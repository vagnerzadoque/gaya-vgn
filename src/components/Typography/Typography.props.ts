export interface TypographyProps {
  /**
   * Specify an optional className to be added to your Typography
   */
  className?: string

  /**
   * Optional ID for testing
   */
  testID?: string;

  /**
   * Specify the variant of Typography you want to create.
   * If no variant is specified, a 'body1' will be rendered.
   */
  variant?:
    'heading1' |
    'heading2' |
    'heading3' |
    'heading4' |
    'heading5' |
    'heading6' |
    'subtitle1' |
    'subtitle2' |
    'body1' |
    'body2' |
    'caption' |
    'overline',

  /**
   * Children to be rendered inside the Typography.
   * Should be pass a string like e.g. Lorem Ipsum
   * @required
   */
   children: React.ReactNode;
}
