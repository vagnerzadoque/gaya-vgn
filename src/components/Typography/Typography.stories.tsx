import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { Typography, TypographyProps } from '.'

const componentStatus = `
> used to control and display text styles consistently across an application

## Properties
| Property                      | Values                                                 |    Status           |
|---                            |                                                     ---|                  ---|
| **variant**         | heading1, heading2, heading3, heading4, heading5, <br /> heading6, subtitle1, subtitle2, body1, body2, caption, overline  | âœ… Available        |

## Technical Usages Examples
`

const meta: Meta<TypographyProps> = {
  title: 'Components/Typography',
  component: Typography,
  parameters: {
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<TypographyProps>;

export const Playground: Story = {
  args: {
    children: "Hello world"
  },
  render: (args) => <Typography {...args} />
};
