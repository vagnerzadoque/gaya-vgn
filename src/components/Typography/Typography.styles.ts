/* eslint-disable complexity */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'

type TypographyVariant = {
  variant: 'heading1' |
  'heading2' |
  'heading3' |
  'heading4' |
  'heading5' |
  'heading6' |
  'subtitle1' |
  'subtitle2' |
  'body1' |
  'body2' |
  'caption' |
  'overline'
}

const setFontFamily = (theme: Theme) => ({ variant }: TypographyVariant) => {
  if (variant.includes('heading')) {
    return `"${theme.typography.headline.fontFamily}"`
  }

  return `"${theme.typography.body.regular.fontFamily}"`
}

const setTextTransform = () => ({ variant }: TypographyVariant) => {
  if (variant === 'overline') {
    return 'uppercase'
  }

  return 'none'
}

const styles = createUseStyles((theme: Theme) => ({
  text: {
    fontFamily: setFontFamily(theme),
    fontWeight: ({ variant }: TypographyVariant) => theme[variant].fontWeight,
    fontSize: ({ variant }: TypographyVariant) => theme[variant].fontSize,
    lineHeight: ({ variant }: TypographyVariant) => theme[variant].lineHeight,
    letterSpacing: ({ variant }: TypographyVariant) => theme[variant].letterSpacing,
    textTransform: setTextTransform(),
    margin: 0
  }

}))

export default styles
