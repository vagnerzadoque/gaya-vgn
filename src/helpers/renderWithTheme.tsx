import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
