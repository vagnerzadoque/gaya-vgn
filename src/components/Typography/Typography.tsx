import React, { useEffect, useState } from 'react'
import styles from './Typography.styles'
import { TypographyProps } from './Typography.props'

const Typography = React.forwardRef<HTMLDivElement, TypographyProps>(
  (props, ref) => {
    const {
      variant = 'body1',
      ...rest
    } = props
    const { text } = styles({ variant })
    const [CustomTag, setCustomTag] = useState('p' as keyof JSX.IntrinsicElements)

    useEffect(() => {
      const suffix = variant[variant.length - 1]

      let nameElement

      if (variant.includes('heading')) {
        nameElement = `h${suffix}`
      } else if (variant === 'caption' || variant === 'overline') {
        nameElement = 'span'
      } else {
        nameElement = 'p'
      }

      setCustomTag(nameElement as keyof JSX.IntrinsicElements)
    }, [variant])

    return (
      <div ref={ref} {...rest}>
        <CustomTag className={`${text} Typography-gaya`}>{props.children}</CustomTag>
      </div>
    )
  }
)

export default Typography
