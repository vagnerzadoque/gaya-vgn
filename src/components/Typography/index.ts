import Typography from './Typography'

export { Typography }
export { TypographyProps } from './Typography.props'
export default Typography
