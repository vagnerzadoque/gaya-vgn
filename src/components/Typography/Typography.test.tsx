import React from 'react'
import Typography from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

describe('Typography component', () => {
  it('should render the component as body1 when variant is not defined', () => {
    const { styles, component } = renderWithTheme(<Typography>hello</Typography>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
    expect(component.getByText(/hello/i)).toBeInTheDocument()
    expect(styles.includes('font-size: 16px;')).toBeTruthy()
  })

  describe('Heading variants', () => {
    it('should render the component as heading1', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading1">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 96px;')).toBeTruthy()
    })

    it('should render the component as heading2', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading2">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 64px;')).toBeTruthy()
    })

    it('should render the component as heading3', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading3">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 48px;')).toBeTruthy()
    })

    it('should render the component as heading4', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading4">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 32px;')).toBeTruthy()
    })

    it('should render the component as heading5', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading5">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 24px;')).toBeTruthy()
    })

    it('should render the component as heading6', () => {
      const { styles, component } = renderWithTheme(<Typography variant="heading6">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByRole('heading')).toBeInTheDocument()
      expect(styles.includes('font-size: 20px;')).toBeTruthy()
    })
  })

  describe('Subtitle variants', () => {
    it('should render the component as subtitle1', () => {
      const { styles, component } = renderWithTheme(<Typography variant="subtitle1">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 16px;')).toBeTruthy()
    })

    it('should render the component as subtitle2', () => {
      const { styles, component } = renderWithTheme(<Typography variant="subtitle2">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 14px;')).toBeTruthy()
    })
  })

  describe('Body variants', () => {
    it('should render the component as body1', () => {
      const { styles, component } = renderWithTheme(<Typography variant="body1">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 16px;')).toBeTruthy()
    })

    it('should render the component as body2', () => {
      const { styles, component } = renderWithTheme(<Typography variant="body2">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 14px;')).toBeTruthy()
    })
  })

  describe('Caption variant', () => {
    it('should render the component as caption', () => {
      const { styles, component } = renderWithTheme(<Typography variant="caption">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 12px;')).toBeTruthy()
    })
  })

  describe('Overline variant', () => {
    it('should render the component as overline', () => {
      const { styles, component } = renderWithTheme(<Typography variant="overline">hello</Typography>)

      expect([styles.toString(), component.container]).toMatchSnapshot()
      expect(component.getByText(/hello/i)).toBeInTheDocument()
      expect(styles.includes('font-size: 12px;')).toBeTruthy()
      expect(styles.includes('text-transform: uppercase;')).toBeTruthy()
    })
  })
})
