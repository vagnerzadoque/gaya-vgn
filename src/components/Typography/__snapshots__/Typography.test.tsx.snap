// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component Body variants should render the component as body1 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 16px;
  line-height: 1.5;
  letter-spacing: 0.51px;
  text-transform: none;
}",
  <div>
    <div>
      <p
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </p>
    </div>
  </div>,
]
`;

exports[`Typography component Body variants should render the component as body2 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 14px;
  line-height: 1.5;
  letter-spacing: 0.22px;
  text-transform: none;
}",
  <div>
    <div>
      <p
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </p>
    </div>
  </div>,
]
`;

exports[`Typography component Caption variant should render the component as caption 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 12px;
  line-height: 1.5;
  letter-spacing: 0.38px;
  text-transform: none;
}",
  <div>
    <div>
      <span
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </span>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading1 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 96px;
  line-height: 1.5;
  letter-spacing: 0;
  text-transform: none;
}",
  <div>
    <div>
      <h1
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h1>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading2 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 64px;
  line-height: 1.5;
  letter-spacing: 0;
  text-transform: none;
}",
  <div>
    <div>
      <h2
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h2>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading3 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 48px;
  line-height: 1.5;
  letter-spacing: 0;
  text-transform: none;
}",
  <div>
    <div>
      <h3
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h3>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading4 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 32px;
  line-height: 1.5;
  letter-spacing: 0;
  text-transform: none;
}",
  <div>
    <div>
      <h4
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h4>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading5 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 24px;
  line-height: 1.5;
  letter-spacing: 0;
  text-transform: none;
}",
  <div>
    <div>
      <h5
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h5>
    </div>
  </div>,
]
`;

exports[`Typography component Heading variants should render the component as heading6 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 500;
  font-size: 20px;
  line-height: 1.5;
  letter-spacing: 0.24px;
  text-transform: none;
}",
  <div>
    <div>
      <h6
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </h6>
    </div>
  </div>,
]
`;

exports[`Typography component Overline variant should render the component as overline 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 500;
  font-size: 12px;
  line-height: 1.5;
  letter-spacing: 1.92px;
  text-transform: uppercase;
}",
  <div>
    <div>
      <span
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </span>
    </div>
  </div>,
]
`;

exports[`Typography component Subtitle variants should render the component as subtitle1 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 500;
  font-size: 16px;
  line-height: 1.5;
  letter-spacing: 0.13px;
  text-transform: none;
}",
  <div>
    <div>
      <p
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </p>
    </div>
  </div>,
]
`;

exports[`Typography component Subtitle variants should render the component as subtitle2 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 500;
  font-size: 14px;
  line-height: 1.5;
  letter-spacing: 0.11px;
  text-transform: none;
}",
  <div>
    <div>
      <p
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </p>
    </div>
  </div>,
]
`;

exports[`Typography component should render the component as body1 when variant is not defined 1`] = `
[
  ".text-0-2-1 {
  margin: 0;
}
.text-d0-0-2-2 {
  font-family: "Roboto";
  font-weight: 400;
  font-size: 16px;
  line-height: 1.5;
  letter-spacing: 0.51px;
  text-transform: none;
}",
  <div>
    <div>
      <p
        class="text-0-2-1 text-d0-0-2-2 Typography-gaya"
      >
        hello
      </p>
    </div>
  </div>,
]
`;
